Editorial updates to navigating merklelogs (#876)
datatrails/datatrails-docs: content/developers/developer-patterns/navigating-merklelogs/index.md

The idtimestamp of the last leaf entry added to the log is always set in the header field.

> In the hex data above, the idtimestamp of the last entry in the log is `8e84dbbb6513a6`, the version is `0`, the timestamp epoch is `1`, the massif height is `14`, and the massif index is `0`.

Author: @robinbryce, should `8e84dbbb6513a6` be visible in [line 131](https://github.com/datatrails/datatrails-docs/pull/876/files#diff-cd9ba8ebeb5306314bf53c68a009c107b6699df01c0d37a89e52b93233296517R131)

Reviewer: The diff is a little confusing, but `8e84dbbb6513a6` should be visible in the example output from the curl with the range header `-H "Range: bytes=0-31"`. If you've tried that comand and not obtained the original example

00000000000000008e84dbbb6513a60000000000000000000000010e00000000

Then something is wrong. uploading the .log file to https://hexed.it/ and have a look

Author: We'll fix once merged into your branch. This fork is making this complex to review.

  {{< /tab >}}
{{< /tabs >}}

In this example, the last entry in the log (at that time) was `2024/03/28`, a little after 11.30am.

> ## Triedata Entries Are 512 Bytes Each and Are Formed From Two Fields

Reviewer: ```suggestion
## Trie data Entries Are 512 Bytes Each and Are Formed From Two Fields
```

lets just use the words "trie data" if the camel case convention is distracting

The idtimestamp is a hex string and prefixed with `01` which is the epoch from the header.

To condition the string value, strip the leading `01` and convert the remaining hex to binary.
Then substitute those bytes, in presentation order, for idTimestamp above.

Reworking the python example above to deal with the epoch prefix would look like this:

{{< tabs name="convert idtimestamp" >}}
   {{< tab name="Python" >}}

  ```python
  epoch = 1
  unixms=int((
    bytes.fromhex("018e84dbbb6513a6"[2:])[:-2]).hex(), base=16
    ) + epoch*((2**40)-1)
  ```

  {{< /tab >}}
{{< /tabs >}}

The bytes for the hash are:  
`bytes.fromhex("018e84dbbb6513a6"[2:])`

Given an event identity of:  
  `assets/31de2eb6-de4f-4e5a-9635-38f7cd5a0fc8/events/21d55b73-b4bc-4098-baf7-336ddee4f2f2`

The trieKey would be:  
`b8e04443d64b8f1603fff250952b29e71d6c2d221afd8d60dec133c63e7325d9`

The following python snippet generates a trieKey from the event data to confirm what should be in the index at a specific position.

{{< tabs name="convert idtimestamp" >}}
   {{< tab name="Python" >}}
  ```python
  h = hashlib.sha256()
  h.update(bytes([0]))
  h.update(bytes.fromhex("018e84dbbb6513a6"[2:]))

>   h.update("assets/31de2eb6-de4f-4e5a-9635-38f7cd5a0fc8/events/21d55b73-b4bc-4098-baf7-336ddee4f2f2".encode())

Author: Should this snippet work? Must we keep the IDs, or redact them `xxxxxx-xxx-xxx-xxxxxxxxxxxx`

Reviewer: This snipet should work. I very much wanted the property that the examples could be confirmed from real log data. And I believe that is the case today

Author: Lets merge this into your branch, and iterate from there.

For the forseeable future (at least months) DataTrails does not anticipate needing to reconfigure the massif height.
{{< /note >}}

## Reading a Specific MMR Node by mmrindex

To read a specific mmr node, find the smallest `Last Node` in [Massif Blob Pre-Calculated Offsets](/developers/developer-patterns/massif-blob-offset-tables) that is greater than your *mmrIndex* and use that row as your massif index.

Taking the massif index of 0 (row 0) use the first mmrIndex:

{{< tabs name="convert idtimestamp" >}}
   {{< tab name="Python" >}}
  ```python
  LOGSTART=1048864
  MMRINDEX=0
  curl -s \
  -H "Range: bytes=$(($LOGSTART+$MMRINDEX*32))-$(($LOGSTART+$MMRINDEX*32+31))" \
  -H "x-ms-blob-type: BlockBlob" \
  -H "x-ms-version: 2019-12-12" \
  https://jitavidfd1103b1099ab3aa.blob.core.windows.net/\
  merklelogs/v1/mmrs/tenant/73b06b4e-504e-4d31-9fd9-5e606f329b51/0/massifs/\
  0000000000000000.log  | od -An -tx1 | tr -d ' \n'

  a45e21c14ee5a0d12d4544524582b5feb074650e6bb2b31ed9a3aeffe4883099
  ```
  {{< /tab >}}
{{< /tabs >}}

Optionally use [veracity](https://github.com/datatrails/veracity/) to confirm:

{{< tabs name="verfify with veracity" >}}
   {{< tab name="bash" >}}
  ```bash

>   veracity -s jitavidfd1103b1099ab3aa \

Author: Need to verify this command works, and should it work on a specific tenant, or a variable provided by the user?

Reviewer: also, veracity no longer requires the fussy `-s ACCOUNT` flag, its just `--data-url` which defaults to prod anyway

Author: We'll cleanup in your branch

If you know the byte offset in the blob for the start of the mmr data then you can check the number of mmr nodes currently in it by doing `(blobSize - mmrDataStart)/32`.

> ## The Peak Stack and Mmr Data Sizes Are Computable

Reviewer: ```suggestion
## The Peak Stack and MMR Data Sizes Are Computable
```

MMR is an acronym

If the log is re-configured having first been available at:  
`https://app.datatrails.ai/verifiabledata/merklelogs/v1/mmrs/tenant/72dc8b10-dde5-43fe-99bc-16a37fd98c6a/0/`

On re-configuration the log will become available (without downtime) at:  
`https://app.datatrails.ai/verifiabledata/merklelogs/v1/mmrs/tenant/72dc8b10-dde5-43fe-99bc-16a37fd98c6a/1/`

The previous path will no longer receive any additions.

{{< note >}}
For the forseeable future (at least months) DataTrails does not anticipate needing to reconfigure the massif height.
{{< /note >}}

## Reading a Specific MMR Node by mmrindex

To read a specific mmr node, find the smallest `Last Node` in [Massif Blob Pre-Calculated Offsets](/developers/developer-patterns/massif-blob-offset-tables) that is greater than your *mmrIndex* and use that row as your massif index.

Taking the massif index of 0 (row 0) use the first mmrIndex:

{{< tabs >}}
   {{< tab name="Python" >}}

  ```python
  LOGSTART=1048864
  MMRINDEX=0
  curl -s \
  -H "Range: bytes=$(($LOGSTART+$MMRINDEX*32))-$(($LOGSTART+$MMRINDEX*32+31))" \
  -H "x-ms-blob-type: BlockBlob" \
  -H "x-ms-version: 2019-12-12" \

>   https://jitavidfd1103b1099ab3aa.blob.core.windows.net/\

Reviewer: Ah, because we couldn't affort the max theoretical rate of completely public access all of these urls ned to change. We need. to replace `https://jitavidfd1103b1099ab3aa.blob.core.windows.net/` with the proxy url `https://app.datatrails.ai/verifiabledata/`

  ```

  {{< /tab >}}
{{< /tabs >}}

The example javascript in [Massif Blob Pre-Calculated Offsets](/developers/developer-patterns/massif-blob-offset-tables#the-algorithms-backing-the-table-generation) can be used to accomplish this computationally.

## Which Nodes

Typically, verifiers would verify the inclusion of an event in the log.
The inclusion is verified by selecting the sibling path needed to recreate the root hash starting from your leaf hash.
Create the leaf hash using the original pre-image data of your event and the *commit* values assigned to it when it was included in the log.

For example:

* The V3 canonical set of fields from your event
* The `merklelog_entry.commit.index` (the mmrIndex of the event in the log)
* The `merklelog_entry.commit.idtimestamp` uniquely placing the record of the log in time (according to our cloud service provider)

Determining the sibling path will be a future article.
First, set the scene by covering how the logical tree nodes map to storage.

What is a sibling path?
To understand this we need to dig into how DataTrails organizes the nodes in the Merkle log in storage and memory.

> ## Tree Mapping to Storage

Reviewer: ```suggestion
## Tree Layout in Storage
```

This means that when producing the path we need to access nodes throughout the tree to produce the proof.

Using the "canonical" MMR log for illustration, we get this diagram:

```output
                          We call these the 'spur' nodes,
                       14 as they each depend on ancestor nodes
                          \        affectionately called the alpine zone
                6           13             22   
                  \            \              \
    h=2 1 |  2  |  5  |   9  |  12   |  17   | 21  | -- massif 'tree line'
          |     |     |      |       |       |     |
        0 |0   1| 3  4| 7   8|10   11|15   16|18 19|
```

The sibling *proof* path for the leaf with `mmrIndex` `7` would be [`8`, `12`, `6`], and the "peak bagging" algorithm would then be applied to get the root.

A very nice visualization of how the peaks combine is available in this paper on
[cryptographic, asynchronous, accumulators](https://eprint.iacr.org/2015/718.pdf) (see Fig 4, page 12)

## Visualizing the “Look Back” Peak Stack

A specific challenge for log implementations with very large data sets is answering "how far back" or "how far forward" may be seen?

> MMRs differ from classic binary Merkle trees in how the incomplete sub-trees are combined into a common root.

Reviewer: ```suggestion
MMRs differ from classic binary Merkle trees in how the incomplete sub-trees are combined into a common root. Further, algorithms for proofs of inclusion and consistency that work directly with the peaks offer distinct advantages for relying parties that wish to efficiently check that, once added to a log, an entry continues to be maintained in the log as it evolves```
